src/jdk.incubator.code/share/classes/jdk/incubator/code/dialect/core/CoreOp.java

```diff
@@ -318,6 +320,85 @@ public Block.Builder lower(Block.Builder b, CodeTransformer _ignore) {
             b.rebind(b.context(), CodeTransformer.LOWERING_TRANSFORMER).op(this);
             return b;
         }
+
+        static CoreOp.FuncOp convertInvokeToFuncCallOp(CoreOp.FuncOp funcOp, MethodHandles.Lookup l) {
+            return funcOp.transform(funcOp.funcName(), (blockBuilder, op) -> {
+                if (op instanceof JavaOp.InvokeOp iop) {
+                    Method invokeOpCalledMethod = null;
+                    try {
+                        invokeOpCalledMethod = iop.invokeDescriptor().resolveToMethod(l);
+                    } catch (ReflectiveOperationException e) {
+                        throw new RuntimeException(e);
+                    }
+                    if (invokeOpCalledMethod instanceof Method m) {
+                        CoreOp.FuncOp f = Op.ofMethod(m).orElse(null);
+                        if (f != null) {
+                            Op.Result result = blockBuilder.op(CoreOp.funcCall(
+                                    iop.invokeDescriptor().name(),
+                                    f.invokableType(),
+                                    blockBuilder.context().getValues(iop.operands())));
+                            blockBuilder.context().mapValue(op.result(), result);
+                            return blockBuilder;
+                        }
+                    }
+                }
+                blockBuilder.op(op);
+                return blockBuilder;
+            });
+        }
+
+        static CoreOp.FuncOp invokeToFuncOp(JavaOp.InvokeOp invokeOp, MethodHandles.Lookup l) {
+            try {
+                Method method = invokeOp.invokeDescriptor().resolveToMethod(l);
+                return Op.ofMethod(method).orElse(null);
+            } catch (ReflectiveOperationException e) {
+                throw new IllegalStateException("Could not resolve invokeOp to method");
+            }
+        }
+
+        /**
+         * Returns a moduleOp with the given funcOp as the root.
+         * The funcOps in the moduleOp functionTable are returned in the order encountered within the funcOp.
+         */
+        public static CoreOp.ModuleOp ofFuncOp(CoreOp.FuncOp f, MethodHandles.Lookup l) {
+            Set<CoreOp.FuncOp> visited = new HashSet<>();
+            List<CoreOp.FuncOp> funcs = new ArrayList<>();
+            Deque<CoreOp.FuncOp> stack = new LinkedList<>();
+
+            stack.add(f);
+            while (!stack.isEmpty()) {
+                CoreOp.FuncOp cur = stack.removeLast();
+                if (!visited.add(cur)) continue;
+                funcs.add(convertInvokeToFuncCallOp(cur, l));
```

Comment:
You are traversing the model and resolving methods twice, you can combine the two. If you want to keep it a separate method then pass the temporary list of the encountered invokes to reflectable methods.